### PR TITLE
[14.0] sale_stock_available_to_promise_release: improve '_get_next_replenishment_date' perf

### PR DESCRIPTION
**Pull request by sebalix**
Checking ongoing states instead of finished ones (done and canceled) helps PostgreSQL to perform the filter on far less lines (done lines are increasing a lot over time), speeding up the query to get the next replenishment date.

Some figures based on a product having ~82k moves over 5M in the table. The query has been improved by a factor of 50.

**Before**
```sql
 Limit  (cost=1000.49..36535.12 rows=1 width=12) (actual time=1233.129..1236.150 rows=1 loops=1)
   ->  Nested Loop  (cost=1000.49..320812.16 rows=9 width=12) (actual time=1233.128..1236.148 rows=1 loops=1)
         Join Filter: (stock_move.picking_type_id = stock_picking_type.id)
         Rows Removed by Join Filter: 7
         ->  Gather Merge  (cost=1000.49..320786.59 rows=162 width=16) (actual time=1232.167..1236.087 rows=2 loops=1)
               Workers Planned: 4
               Workers Launched: 2
               ->  Parallel Index Scan using stock_move_date_index on stock_move  (cost=0.43..319767.24 rows=40 width=16) (actual time=860.442..1226.830 rows=3 loops=3)
                     Filter: ((((state)::text <> ALL ('{cancel,done}'::text[])) OR (state IS NULL)) AND (product_id = 7449) AND (company_id = 1))
                     Rows Removed by Filter: 1731876
         ->  Materialize  (cost=0.00..6.25 rows=8 width=4) (actual time=0.011..0.027 rows=4 loops=2)
               ->  Seq Scan on stock_picking_type  (cost=0.00..6.21 rows=8 width=4) (actual time=0.018..0.048 rows=6 loops=1)
                     Filter: ((company_id = 1) AND ((code)::text = 'incoming'::text))
                     Rows Removed by Filter: 104
 Planning Time: 0.496 ms
 Execution Time: 1236.199 ms
```

**After**
```sql
 Limit  (cost=6776.42..6776.42 rows=1 width=12) (actual time=25.782..25.785 rows=1 loops=1)
   ->  Sort  (cost=6776.42..6776.44 rows=9 width=12) (actual time=25.780..25.782 rows=1 loops=1)
         Sort Key: stock_move.date
         Sort Method: top-N heapsort  Memory: 25kB
         ->  Nested Loop  (cost=6569.58..6776.37 rows=9 width=12) (actual time=25.627..25.770 rows=6 loops=1)
               Join Filter: (stock_move.picking_type_id = stock_picking_type.id)
               Rows Removed by Join Filter: 681
               ->  Bitmap Heap Scan on stock_move  (cost=6569.58..6750.80 rows=162 width=16) (actual time=25.390..25.522 rows=116 loops=1)
                     Recheck Cond: (((product_id = 7449) AND (company_id = 1) AND ((state)::text = ANY ('{partially_available,confirmed,assigned,waiting,draft}'::text[]))) OR (state IS NULL))
                     Filter: ((product_id = 7449) AND (company_id = 1))
                     Heap Blocks: exact=104
                     ->  BitmapOr  (cost=6569.58..6569.58 rows=162 width=0) (actual time=25.367..25.368 rows=0 loops=1)
                           ->  Bitmap Index Scan on stock_move_product_location_index  (cost=0.00..6567.96 rows=162 width=0) (actual time=25.356..25.356 rows=125 loops=1)
                                 Index Cond: ((product_id = 7449) AND (company_id = 1) AND ((state)::text = ANY ('{partially_available,confirmed,assigned,waiting,draft}'::text[])))
                           ->  Bitmap Index Scan on stock_move_state_index  (cost=0.00..1.54 rows=1 width=0) (actual time=0.008..0.008 rows=0 loops=1)
                                 Index Cond: (state IS NULL)
               ->  Materialize  (cost=0.00..6.25 rows=8 width=4) (actual time=0.000..0.001 rows=6 loops=116)
                     ->  Seq Scan on stock_picking_type  (cost=0.00..6.21 rows=8 width=4) (actual time=0.028..0.057 rows=6 loops=1)
                           Filter: ((company_id = 1) AND ((code)::text = 'incoming'::text))
                           Rows Removed by Filter: 104
 Planning Time: 0.647 ms
 Execution Time: 25.862 ms
```

### CODE REVIEW

**Review by TDu — APPROVED** [review]

File: sale_stock_available_to_promise_release/models/product_product.py [hunk 1]
```diff
@@ -12,7 +12,17 @@ def _get_next_replenishment_date(self):
...
+                (
+                    "state",
+                    "in",
+                    [
+                        "draft",
+                        "waiting",
+                        "confirmed",
+                        "partially_available",
+                        "assigned",
+                    ],
```
Comment (simahawk) [inline sale_stock_available_to_promise_release/models/product_product.py]:
```suggestion
                    (
                        "draft",
                        "waiting",
                        "confirmed",
                        "partially_available",
                        "assigned",
                    ),
```

**Review by simahawk — APPROVED** [review]
minor remark

File: sale_stock_available_to_promise_release/models/product_product.py (hunk 1 above)
Comment (sebalix) [reply sale_stock_available_to_promise_release/models/product_product.py]: updated

**Review by jbaudoux — COMMENTED** [review]
I tried and I have opposite result
Test DB with 10 million stock moves
Postgresql 16

**Without limit 1 - Before:**
 Execution Time: 3028.400 ms
**Without limit 1 - After:**
 Execution Time: 67.464 ms

**With limit 1 - Before:**
 Execution Time: 0.320 ms
**With limit 1 - After:**
 Execution Time: 41.246 ms

**Comment by sebalix** [comment]
@jbaudoux is your test still returning these results? Or did you keep the "not in" operator instead of "in" by mistake when testing it (as discussed in private)?

**Comment by simahawk** [comment]
@jbaudoux ping

**Comment by jbaudoux** [comment]
I forgot the `order by date asc`. Here are my results:

### state not in
```
explain analyze select id from stock_move where picking_type_id in (select id from stock_picking_type where code = 'incoming') and state not in ('done', 'cancel') and product_id = 3224 order by date asc limit 1;
                                                                            QUERY PLAN                                                                             
-------------------------------------------------------------------------------------------------------------------------------------------------------------------
 Limit  (cost=4521.39..4521.39 rows=1 width=12) (actual time=19.853..19.855 rows=0 loops=1)
   ->  Sort  (cost=4521.39..4521.39 rows=1 width=12) (actual time=19.849..19.851 rows=0 loops=1)
         Sort Key: stock_move.date
         Sort Method: quicksort  Memory: 25kB
         ->  Nested Loop  (cost=17.19..4521.38 rows=1 width=12) (actual time=19.835..19.837 rows=0 loops=1)
               Join Filter: (stock_move.picking_type_id = stock_picking_type.id)
               Rows Removed by Join Filter: 28
               ->  Seq Scan on stock_picking_type  (cost=0.00..11.75 rows=1 width=4) (actual time=0.035..0.050 rows=4 loops=1)
                     Filter: ((code)::text = 'incoming'::text)
                     Rows Removed by Filter: 38
               ->  Bitmap Heap Scan on stock_move  (cost=17.19..4509.60 rows=2 width=16) (actual time=2.363..4.941 rows=7 loops=4)
                     Recheck Cond: (product_id = 3224)
                     Filter: ((state)::text <> ALL ('{done,cancel}'::text[]))
                     Rows Removed by Filter: 2974
                     Heap Blocks: exact=11624
                     ->  Bitmap Index Scan on stock_move_product_location_index  (cost=0.00..17.19 rows=1167 width=0) (actual time=0.654..0.654 rows=2981 loops=4)
                           Index Cond: (product_id = 3224)
 Planning Time: 1.609 ms
 Execution Time: 20.028 ms
```
### state in
```
explain analyze select id from stock_move where picking_type_id in (select id from stock_picking_type where code = 'incoming') and state in ('partially_available','confirmed','assigned','waiting','draft') and product_id = 3224 order by date asc limit 1;
                                                                            QUERY PLAN                                                                             
-------------------------------------------------------------------------------------------------------------------------------------------------------------------
 Limit  (cost=124.34..124.34 rows=1 width=12) (actual time=1.778..1.783 rows=0 loops=1)
   ->  Sort  (cost=124.34..124.34 rows=1 width=12) (actual time=1.774..1.778 rows=0 loops=1)
         Sort Key: stock_move.date
         Sort Method: quicksort  Memory: 25kB
         ->  Nested Loop  (cost=0.43..124.33 rows=1 width=12) (actual time=1.757..1.760 rows=0 loops=1)
               Join Filter: (stock_move.picking_type_id = stock_picking_type.id)
               Rows Removed by Join Filter: 28
               ->  Seq Scan on stock_picking_type  (cost=0.00..11.75 rows=1 width=4) (actual time=0.039..0.089 rows=4 loops=1)
                     Filter: ((code)::text = 'incoming'::text)
                     Rows Removed by Filter: 38
               ->  Index Scan using stock_move_product_location_index on stock_move  (cost=0.43..112.54 rows=3 width=16) (actual time=0.036..0.401 rows=7 loops=4)
                     Index Cond: ((product_id = 3224) AND ((state)::text = ANY ('{partially_available,confirmed,assigned,waiting,draft}'::text[])))
 Planning Time: 1.551 ms
 Execution Time: 1.954 ms

**Review by jbaudoux — APPROVED** [review]

**Pull request merged**